### DISCUSSION TITLE
tutorials/lift-off-part3/08-the-usequery-hook-with-variables

### DISCUSSION THREAD

**Discussion by giscus** [apollographql-education/content-feedback#27] [course feedback]
# tutorials/lift-off-part3/08-the-usequery-hook-with-variables

https://www.apollographql.com/tutorials/lift-off-part3/08-the-usequery-hook-with-variables

**Comment by pawangujral** [comment]
Shouldn't this be `numberOfModules` instead of `numberOfViews` in Query. Maybe a typo

**Reply by mabuyo** [reply]
Hey @pawangujral , sorry we missed your comment! It _should_ be `numberOfViews`, that is a valid field in our schema. The number of modules is covered by the `modulesCount` field. Hope that helps!

**Comment by macgeargear** [comment]
how to fix this error - ERROR: request to https://odyssey-lift-off-rest-api.herokuapp.com/tracks failed, reason: Socket connection timeout

**Reply by mabuyo** [reply]
Hey there @macgeargear - sorry to hear you're running into issues. I checked the endpoint and confirmed that it's working. Where/when are you seeing the error? Happy to also look at your code and try to reproduce the issue. Let me know!

**Comment by lst2018** [comment]
hi, 
I got an error:
the result of http://localhost:3000/track/c_0
ERROR: Cannot return null for non-nullable field Module.id.
could anybody help?

**Reply by lizhennessy** [reply]
Hi there @lst2018 ! I'm happy to take a look. Do you have a link to your repository that I could access to see what might be going wrong?

**Reply by gijs-martens** [reply]
I had the same thing, restarting the server resolved it for me.
I think it might be the difference between running `npm start` and `npm run dev`, a discrepancy in the tutorial.

**Comment by lst2018** [comment]
hi there @lizhennessy, thanks for your time to have a look at my problem.
I saved the files in the git folder: https://github.com/lst2018/apollographql.git

**Comment by lst2018** [comment]
hi,
I resolve my problem by repeat again this course from the beginning. please ignore my problem, sorry for my question, thank you.

**Comment by lst2018** [comment]
hi all,

anybody find the answer for the question: == Do you see the track page when you change the URL in the browser? What is the first module title on the list of modules when you go to http://localhost:3000/track/c_2? ==
i didn't find the answer in that page.   but i find it if i go the api page: ../track/{id}/modules  by try with track ID 0_2
 
i my side the list modules : this page http://localhost:3000/track/c_2? shown as below: 
am i in the right track on this?

.... 
Modules
Cat-stronomy, an introduction 39m
Famous Catstronauts 31m
Kitty space suit, all you need to know 30m
Cat-strophysics, master class 49m
Crew Dragon, kitty edition 36m
Solar System, crash track 48m
Mars-A-CAT, for dummies 24m
Cat-strodynamics 101 10m
Rover Driving, for beginners 41m
To The Moon, 7 steps program 14m
Space CATastrophes, pro level 25m
Kitty Rockets, the hard way 40m
Space-Cat-Walk, best practices 41m

**Reply by lst2018** [reply]
hi there @mabuyo,  could you please help me

**Reply by lizhennessy** [reply]
Hi there @lst2018 ! You're on the right track - when you go to `http://localhost:3000/track/c_2`, the `Track` itself should be called `Kitty space suit, all you need to know` - just below the course information such as author, views and number of modules, you should see a list of the modules that the course includes! Please let me know if that helps you answer the quiz question!

<img width="991" alt="image" src="https://github.com/apollographql-education/content-feedback/assets/95302380/d2fa0510-6999-436c-94d4-a605fa4e31fe">

**Reply by lst2018** [reply]
hi 
yes, i got the same picture and called Kitty space suit, all you need to know.  but the list of modules is different:

     Modules
Cat-stronomy, an introduction 39m
Famous Catstronauts 31m
Kitty space suit, all you need to know 30m
Cat-strophysics, master class 49m
Crew Dragon, kitty edition 36m
Solar System, crash track 48m
Mars-A-CAT, for dummies 24m
Cat-strodynamics 101 10m
Rover Driving, for beginners 41m
To The Moon, 7 steps program 14m
Space CATastrophes, pro level 25m
Kitty Rockets, the hard way 40m
Space-Cat-Walk, best practices 41m

so i didn't find the answer for this question:
"Do you see the track page when you change the URL in the browser? What is the first module title on the list of modules when you go to http://localhost:3000/track/c_2?"

because the first module title is "Cat-stronomy, an introduction 39m"

in your side, based on your screen the answer is "Pressurized Atmosphere"

only the list modules is different, why, did i miss something?
if i open http://localhost:3000/track/c_0, the list of modules is also same with c_2

thank you

**Reply by lizhennessy** [reply]
Hi @lst2018 ! I see - I'm happy to take a closer look to see if we missed a step somewhere. Can you link me to your repository?

**Reply by lst2018** [reply]
hi @lizhennessy,

i just find the incorrect function in the resolver, 

 modules: ({id}, _,{dataSources})=>{
      return dataSources.trackAPI.getTracksForHome(id);
    }

i change it with the correct function:

 modules: ({id}, _,{dataSources})=>{
      return dataSources.trackAPI.getTrackModules(id);
    }

now i got the answer.

thank you

**Reply by lizhennessy** [reply]
That's great to hear, @lst2018 ! So glad you were able to find the solution. Happy coding!

**Comment by yogeshminocha** [comment]
After making sure, all the steps clearly followed, on client load , i get the blank screen.. Not sure, what i am missing. as soon as import of .track, this starts.

**Reply by Gator2-0** [reply]
It's most likely because you are not exporting Track.
You need to add this line of code at the end of the Track.js.

export default Track;

**Comment by gregkepler** [comment]
I'm doing the Typescript version and I see a Typescript error for `GET_TRACK` when passed into `useQuery`.

```
const { loading, error, data } = useQuery(GET_TRACK, {
  variables: { trackId },
});
```

There error is:
```
Argument of type 'unknown' is not assignable to parameter of type 'DocumentNode | TypedDocumentNode<any, { trackId: string; }>'.ts(2345)
```

What am I missing? Also, I've copied from the final / client folder and I still see this issue.

**Reply by mabuyo** [reply]
@gregkepler Can you paste in your code for the `GET_TRACK` query? is it using the `gql` function imported from the `"../__generated__"` file? Should look something like this:

```
const GET_TRACK = gql(`
  // our query goes here
)`);
```
Note the parentheses encapsulating the backticks.

**Reply by gregkepler** [reply]
@mabuyo. ~Below is my entire `track.tsx` file~

It turns out I had an older version of node enabled for another project. Once I switch to Node v18, it started working.

Sorry about that! Learn from me, kids!

**Reply by mabuyo** [reply]
Ahh, glad you figured it out! Thanks for letting us know!

**Reply by ced-mos** [reply]
In case others had the same issue but the Node version was already updated. My solution was just to regenerate the graphql types with `npm run generate`, although I thought they were already updated.

**Reply by otlyn** [reply]
Perhaps you had the same issue I had. The error went away when I imported `gql` from `@apollo/client` instead of `__generated__`.

**Reply by Darwish98** [reply]
There is no need to import from @apollo/client. To solve the issue you have to have your server running, then on you client run npm run generate. This will update the __generated__ and the error is gone. Please refer to: https://www.apollographql.com/tutorials/lift-off-part1/09-codegen

**Comment by otlyn** [comment]
I needed to import `gql` from `@apollo/client` rather than the `__generated__` folder to get it to work. 
```
import React from "react";
// import { gql } from "../__generated__";
import { gql, useQuery } from "@apollo/client";
import { Layout, QueryResult } from "../components";
import { useParams } from "react-router-dom";
import TrackDetail from "../components/track-detail";

export const GET_TRACK = gql(`
  query GetTrack($trackId: ID!) {
    track(id: $trackId) {
      id
      title
      author {
        id
        name
        photo
      }
      thumbnail
      length
      modulesCount
      numberOfViews
      modules {
        id
        title
        length
      }
      description
    }
  }
`);

const Track = () => {
  const { trackId = "" } = useParams();
  const { loading, error, data } = useQuery(GET_TRACK, {
    variables: { trackId },
  });
  return (
    <Layout>
      <QueryResult error={error} loading={loading} data={data}>
        <TrackDetail track={data?.track} />
      </QueryResult>
    </Layout>
  );
};

export default Track;
```

**Reply by lizhennessy** [reply]
Thanks for sharing the workaround! Curious what version of Node you're on @otlyn ?

**Reply by LeSamerian** [reply]
Had the exact same issue and I also needed to use the same workaround. Node `v22.20.0` is the version I'm using

**Comment by austinav** [comment]
My cat was not interested space cats
